addons/mail_group/models/mail_group.py

```diff
@@ -0,0 +1,319 @@
+                lambda message: message.moderation_status == 'pending_moderation'
+            ))
+
+    @api.depends('member_ids')
+    def _compute_member_count(self):
+        for group in self:
+            group.member_count = len(group.member_ids)
+
+    @api.depends('moderator_ids')
+    @api.depends_context('uid')
+    def _compute_is_moderator(self):
+        for group in self:
+            group.is_moderator = self.env.user.id in group.moderator_ids.ids
+
+    @api.depends('moderation_ids')
+    def _compute_moderation_count(self):
+        for group in self:
+            group.moderation_count = len(group.moderation_ids)
+
+    def _alias_get_creation_values(self):
+        """Return the default values for the automatically created alias."""
+        values = super(MailGroup, self)._alias_get_creation_values()
+        values['alias_model_id'] = self.env['ir.model']._get('mail.group').id
+        values['alias_force_thread_id'] = self.id
+        values['alias_defaults'] = literal_eval(self.alias_defaults or '{}')
+        return values
+
+    @api.onchange('moderation')
+    def _onchange_moderation(self):
+        if self.moderation and self.env.user not in self.moderator_ids:
+            self.moderator_ids |= self.env.user
+
+    # CONSTRAINTS
+
+    @api.constrains('moderator_ids')
+    def _check_moderator_email(self):
+        if any(not moderator.email for group in self for moderator in group.moderator_ids):
+            raise ValidationError(_('Moderators must have an email address.'))
+
+    @api.constrains('moderation_notify', 'moderation_notify_msg')
+    def _check_moderation_notify(self):
+        if any(group.moderation_notify and not group.moderation_notify_msg for group in self):
+            raise ValidationError(_('The notification message is missing.'))
+
+    @api.constrains('moderation_guidelines', 'moderation_guidelines_msg')
+    def _check_moderation_guidelines(self):
+        if any(group.moderation_guidelines and not group.moderation_guidelines_msg for group in self):
+            raise ValidationError(_('The guidelines description is missing.'))
+
+    @api.constrains('moderator_ids', 'moderation')
+    def _check_moderator_existence(self):
+        if any(not group.moderator_ids for group in self if group.moderation):
+            raise ValidationError(_('Moderated group must have moderators.'))
+
+    @api.model
+    def message_update(self, message_dict=None):
+        """Add the method to make the mail gateway flow work with this model."""
+        return
+
+    def message_post(self, **kwargs):
+        self.ensure_one()
+        # First create the <mail.message>
+        Mailthread = self.env['mail.thread']
+        values = dict((key, val) for key, val in kwargs.items() if key in self.env['mail.message']._fields)
+        author_id, email_from = Mailthread._message_compute_author(kwargs.get('author_id'), kwargs.get('email_from'), raise_exception=True)
+        values.update({
```

Comment:
Shouldn't we also compute messageID and reply_to in order to have them cleanly set for replies ?

Reply:
We should keep the same message id, otherwise parent might not be found correctly, and the reply_to should be the alias of the mail group, otherwise answer will not create `mail.group.message`

Reply:
Well I meant when posting a message. Indeed through mail gateway message ID will be set, in which case we should keep it. reply-to is set as mail.message create if not set, that's what I had in mind.

I think we should explicitly set them. Global purpose is to remove code in mail.message model and have it at higher level. In short I would take that code from mail.message create and put it here
```
            if not values.get('message_id'):
                values['message_id'] = self._get_message_id(values)
            if 'reply_to' not in values:
                values['reply_to'] = self._get_reply_to(values)
            if 'record_name' not in values and 'default_record_name' not in self.env.context:
                values['record_name'] = self._get_record_name(values)
```